### PR TITLE
Add a sieve implementing filtering based on solver rules

### PR DESCRIPTION
**Pull request by fridex**
## Related Issues and Dependencies

Related: https://github.com/thoth-station/thoth-application/issues/1403
Depends-On: https://github.com/thoth-station/storages/pull/2321

## This introduces a breaking change

- [x] No

### CODE REVIEW

**Comment by fridex** [comment]
/hold

... until new thoth-storages is released and available in this repo.

File: thoth/adviser/sieves/rules.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        super().pre_run()
+
+    def run(self, package_versions: Generator[PackageVersion, None, None]) -> Generator[PackageVersion, None, None]:
+        """Filter out packages that have rules assigned."""
+        for package_version in package_versions:
+            package_tuple = package_version.to_tuple()
+
+            solver_rules = (
+                # Rules specific to index.
+                self.context.graph.get_python_package_version_solver_rules(
```
Comment (pacospace) [inline thoth/adviser/sieves/rules.py:75]: `_all`

File: thoth/adviser/sieves/rules.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+                    package_version.name,
+                    package_version.locked_version,
+                    package_version.index.url,
+                )
+                +
+                # Rules agnostic to index.
+                self.context.graph.get_python_package_version_solver_rules(
```
Comment (pacospace) [inline thoth/adviser/sieves/rules.py:82]: `_all`

File: thoth/adviser/sieves/rules.py (hunk 1 above)
Comment (fridex) [reply thoth/adviser/sieves/rules.py]: Adjusted in 962811f8 PTAL  👍🏻

File: thoth/adviser/sieves/rules.py (hunk 2 above)
Comment (fridex) [reply thoth/adviser/sieves/rules.py]: Adjusted in 962811f8 PTAL  👍🏻

**Review by pacospace — APPROVED** [review]
🍺 thanks lgtm!!

**Pull request merged**